lib/linguist/heuristics.yml

```diff
@@ -412,6 +412,21 @@ disambiguations:
     pattern: '(?i:\bRebol\b)'
   - language: R
     pattern: '<-|^\s*#'
+- extensions: ['.re']
+  rules:
+  - language: Reason
+    pattern:
+    - '^\s*open\s+\w+\s*;$'
+    - '^\s*include\s+\w+\s*;$'
+    - '^\s*module\s+type\s'
+    - '^\s*let\s+\w+:\s+.*=.*;\s*$'
+    - '^\s*let\s+module\s\w+\s*=\s*{'
```

Comment:
This is more optimal (and readable, IMHO):

~~~regexp
^\s*module\s+type\s
^\s*(?:include|open)\s+\w+\s*;$
^\s*let\s+(?:module\s\w+\s*=\s*{|\w+:\s+.*=.*;\s*$)
~~~

Also, unless trailing whitespace after a semicolon and line-break is illegal (which it probably isn't), I'd replace `;$` with `;\s*$`, just to err on the side of caution.